File: 2018/8.py
class Node():
	def __init__(self, numChildren, numMetaData, parentNode):
		self.numChildren = numChildren
		self.numMetaData = numMetaData
		self.metaValue = 0
		self.metaData = []
		self.children = []
		self.parentNode = parentNode

def recur1(ind, nums, currNode, nodes):
	if ind > len(nums):
		return nodes, ind
	for j in range(currNode.numChildren):
		numCC = int(nums[ind])
		numCMD = int(nums[ind+1])
		newNode = Node(numCC, numCMD, currNode)
		currNode.children.append(newNode)
		ind += 2
		nodes.append(newNode)
		nodes, ind = recur1(ind, nums, newNode, nodes)
	for i in range(currNode.numMetaData):
		currNode.metaData.append(int(nums[ind]))
		currNode.metaValue += int(nums[ind])
		ind += 1
	return nodes, ind

def recur2(currNode):
	if currNode.numChildren > 0:
		for child in currNode.children:
			child.metaValue = recur2(child)
		currNode.metaValue = 0
		for datum in currNode.metaData:
			if datum-1 < currNode.numChildren:
				currNode.metaValue += currNode.children[datum-1].metaValue
	return currNode.metaValue
	
def summer(nodes):
	total = 0
	for node in nodes:
		total += node.metaValue
	return total

with open('day8_input.txt', 'r') as infile:
	nums = infile.read().split()

nodes = []
nodes.append(Node(int(nums[0]), int(nums[1]), None))
nodes, length = recur1(2, nums, nodes[0], nodes)
print (summer(nodes))

#Part 2
print (recur2(nodes[0]))

File: 2018/8.rb
def createNode(nums, i)
	numChild = nums[i]; i += 1
	numData = nums[i]; i += 1
	node = [numChild, numData]
	while numChild > 0
		nextChild, i = createNode(nums, i)
		node << nextChild
		numChild -= 1
	end
	while numData > 0
		node << nums[i]
		i += 1
		numData -= 1
	end
	return node, i
end

def metadataCount(node)
	dataStart = 2 + node[0] #Index of last child node + 1 (first metadata)
	total = node[dataStart..].sum #Total of metadata for current node
	for i in 2..dataStart-1
		total += metadataCount(node[i])
	end
	return total
end

def value(node)
	return metadataCount(node) if node[0] == 0 #BASE CASE: NO CHILDREN
	dataStart = 2 + node[0] #Index of last child node + 1 (first metadata)
	value = 0
	node[dataStart..].each do |idx|
		value += value(node[idx+1]) if idx <= node[0]
	end
	return value
end

head, i = createNode(File.open("8.txt").read.split.map(&:to_i), 0)

=begin part 1
	pretty easy, but for some reason the metadata count took me a long time. Creating the node
	structure was easy, reading it wasn't (although it should have been)
=end
p metadataCount(head)

=begin part 2
	Did not take me very long at all, pleased with how I didn't have to change anything about my structure.
=end
p value(head)